File: Cpp/17.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// recursive
class Solution {
public:
    vector<string> digit_to_letters
        = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty()) {
            return {};
        }

        combine("", res, 0, digits);
        return res;
    }

    void combine(string temp, vector<string>& res, int pos, string digits) {
        if (pos >= digits.size()) {
            res.push_back(temp);
            return;
        }

        for (auto c : digit_to_letters[digits[pos] - '0']) {
            combine(temp + c, res, pos + 1, digits);
        }
    }
};

// iterative
class Solution2 {
public:
    vector<string> digit_to_letters
        = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty()) {
            return {};
        }

        res.push_back("");
        for (int i = 0; i < digits.size(); ++i) {
            vector<string> temp;
            for (auto c : digit_to_letters[digits[i] - '0']) {
                for (int i = 0; i < res.size(); ++i) {
                    temp.push_back(res[i] + c);
                }
            }
            res = temp;
        }

        return res;
    }
};

int main() {}

File: Python/17.py
class Solution(object):
    def __init__(self):
        self.mapping = {'0':"", '1':"", '2':"abc", '3':"def", '4':"ghi", '5':"jkl", '6':"mno", '7':"pqrs", '8':"tuv", '9':"wxyz"}
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if digits == "":
            return []
        if len(digits) == 1:
            return list(self.mapping[digits[0]])
        prev = self.letterCombinations(digits[:-1])
        return [s + c for s in prev for c in self.mapping[digits[-1]]]

class Solution(object):
    mapping = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        res = []
        if not digits:
            return []
        self.dfs(digits, "", 0, res)
        return res

    def dfs(self, digits, path, pos, res):
        if pos >= len(digits):
            res.append(path)
            return
        for c in self.mapping[digits[pos]]:
            self.dfs(digits, path+c, pos+1, res)
